ta/pkcs11/src/object.c

```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
 
 	return rc;
 }
+
+static enum pkcs11_rc token_obj_matches_ref(struct obj_attrs *req_attrs,
+					    struct pkcs11_object *obj)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_ObjectHandle hdl = obj->attribs_hdl;
+	TEE_ObjectInfo info = { };
+	struct obj_attrs *attr = NULL;
+	uint32_t read_bytes = 0;
+
+	if (obj->attributes) {
+		if (!attributes_match_reference(obj->attributes, req_attrs))
+			return PKCS11_RV_NOT_FOUND;
+
+		return PKCS11_CKR_OK;
+	}
+
+	if (hdl == TEE_HANDLE_NULL) {
+		res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE,
+					       obj->uuid, sizeof(*obj->uuid),
+					       TEE_DATA_FLAG_ACCESS_READ,
+					       &hdl);
+		if (res) {
+			EMSG("OpenPersistent failed %#"PRIx32, res);
+			return tee2pkcs_error(res);
+		}
+	}
+
+	res = TEE_GetObjectInfo1(hdl, &info);
+	if (res) {
+		EMSG("GetObjectInfo failed %#"PRIx32, res);
+		rc = tee2pkcs_error(res);
+		goto out;
+	}
+
+	attr = TEE_Malloc(info.dataSize, TEE_MALLOC_FILL_ZERO);
+	if (!attr) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	res = TEE_ReadObjectData(hdl, attr, info.dataSize, &read_bytes);
+	if (!res) {
+		res = TEE_SeekObjectData(hdl, 0, TEE_DATA_SEEK_SET);
+		if (res)
+			EMSG("Seek to 0 failed with %#"PRIx32, res);
+	}
+
+	if (res) {
+		rc = tee2pkcs_error(res);
+		EMSG("Read %"PRIu32" bytes, failed %#"PRIx32,
+		     read_bytes, res);
+		goto out;
+	}
+
+	if (read_bytes != info.dataSize) {
+		EMSG("Read %"PRIu32" bytes, expected %"PRIu32,
+		     read_bytes, info.dataSize);
+		rc = PKCS11_CKR_GENERAL_ERROR;
+		goto out;
+	}
+
+	if (!attributes_match_reference(attr, req_attrs)) {
+		rc = PKCS11_RV_NOT_FOUND;
+		goto out;
+	}
+
+	obj->attributes = attr;
+	attr = NULL;
+	obj->attribs_hdl = hdl;
+	hdl = TEE_HANDLE_NULL;
+
+	rc = PKCS11_CKR_OK;
+
+out:
+	TEE_Free(attr);
+	if (obj->attribs_hdl == TEE_HANDLE_NULL && hdl != TEE_HANDLE_NULL)
+		TEE_CloseObject(hdl);
```

Comment:
`TEE_CloseObject()` should be able to handle `TEE_HANDLE_NULL` so I suppose there is no need to have `hdl != TEE_HANDLE_NULL`. Feel free to keep it if it is easier to understand it that way.

`obj->attrib_hdl == TEE_HANDLE_NULL` seems to be here for "re-entry" check that `hdl` does not get freed while in use.

Reply:
agree, `if (obj->attribs_hdl == TEE_HANDLE_NULL)` is enough.

Reply:
Agree, will change in the review patchset